proposals/unixfsv1-dir-as-adl.md

```diff
@@ -0,0 +1,384 @@
+by supplying Unixfsv1 functionality through a clean implementation as an ADL (short for Advanced Data Layout,
+a concept from the IPLD project which is meant to provide a solution area for problems like sharding).
+
+In addition to solving the problem examples above,
+this work will also immediately result in several new features becoming reachable
+(such as IPLD Selectors becoming operable over Unixfsv1 pathing!,
+and build a strong foundation for future work.
+
+We believe that that the work outlined below will
+make the stack more robust,
+more consistent,
+and easier to understand for future developers both inside and nearby the core stack.
+
+### Background & intent
+_Describe the desired state of the world after this project? Why does that matter?_
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives.
+-->
+
+Unixfsv1 is a large and key subsystem of IPFS.
+It is the internal format used by IPFS whenever users are experiencing "files" and "directories".
+(When users are using "dag" APIs or other direct-to-IPLD APIs, they are not (necessarily) experiencing Unixfsv1;
+most other APIs, such as `ipfs add`, `ipfs cat`, including also MFS, etc, _are_ experiencing Unixfsv1.)
+
+Unixfsv1 implementation components are currently scattered among many source code repositories,
+and the API boundaries are not especially clear.  Maintenance is generally difficult.
+It would be desirable for our own development and productivity if more of this implementation was unified in one place.
+
+At the same time, Unixfsv1's current implementation in golang predate much of the latest generation of work in IPLD.
+We believe significant gains of features, as well as clarity of design,
+are possible by rebuilding Unixfsv1 features in the form of an IPLD concept known as ADLs.
+
+Closer to users: The APIs made available to Unixfsv1 from IPFS can be confusing.
+Some very low-level features are made available by IPFS APIs, but are not always safe to use,
+because they may do incorrect things in "corner cases" which are not obvious to end users, resulting in a general air of fragility.
+Some high level features are made available by other IPFS APIs, but jump to such a high level
+that users with specific needs can be observed to jump back down to the low level APIs
+(and then procede to use them incorrectly, because those APIs do not guide the user to good usage).
+
+Example user stories such as "patch this directory" currently don't work out well, because while we offer "patch" APIs in some places,
+they only work on low-level data, or, they work on high level data, but in ways that are not entirely desirable:
+
+- In the low-level way: we encounter problematic scenarios because low-level abstractions do not include enough knowledge
```

Comment:
Can we make clear what the low-level way is?  I assume it's `ipfs object patch add-link`